dali/pipeline/operator/sequence_operator.h

```diff
@@ -0,0 +1,627 @@
+    if (ws.has_stream()) {
+      expanded.set_stream(ws.stream());
+    }
+  }
+
+  template <typename InputBackend>
+  void AddInputHelper(ws_input_t<InputBackend> input) {
+    expanded_.AddInput(input);
+  }
+
+  void AddArgumentInputHelper(const std::string &arg_name,
+                              std::shared_ptr<TensorVector<CPUBackend>> arg_input) {
+    expanded_.AddArgumentInput(arg_name, arg_input);
+  }
+
+  template <typename OutputBackend>
+  void AddOutputHelper(ws_input_t<OutputBackend> output) {
+    expanded_.AddOutput(output);
+  }
+
+  template <typename InputBackend>
+  bool ExpandedInputIsType(int input_idx) const {
+    return expanded_.template InputIsType<InputBackend>(input_idx);
+  }
+
+  template <typename OutputBackend>
+  bool ExpandedOutputIsType(int output_idx) const {
+    return expanded_.template OutputIsType<OutputBackend>(output_idx);
+  }
+
+  template <typename InputBackend>
+  const auto &ExpandedInput(int intput_idx) {
+    return expanded_.template Input<InputBackend>(intput_idx);
+  }
+
+  template <typename OutputBackend>
+  const auto &ExpandedOutput(int output_idx) {
+    return expanded_.template Output<OutputBackend>(output_idx);
+  }
+
+  void ClearExpanded() {
+    expanded_.Clear();
+  }
+
+  TensorVector<CPUBackend> ExpandArgumentLike(const TensorVector<CPUBackend> &arg_tensor,
+                                              const std::string &arg_name, int expand_idx) {
+    const auto &expand_desc = GetInputExpandDesc(expand_idx);
+    DALI_ENFORCE(static_cast<ptrdiff_t>(arg_tensor.num_samples()) == expand_desc.NumSamples(),
+                 make_string("Number of samples passed for argument ", arg_name, " (got ",
+                             arg_tensor.num_samples(),
+                             ") does not match the number of samples in the input (got ",
+                             expand_desc.NumSamples(), ")."));
+    const auto &arg_layout = arg_tensor.GetLayout();
+    int arg_sample_dim = arg_tensor.sample_dim();
+    DALI_ENFORCE(
+        arg_layout.size() == 0 || arg_sample_dim == arg_layout.size(),
+        make_string("Layout of argument input ", arg_name, " has size ", arg_layout.size(),
+                    " which does not match the dimensionality of the argument (got argument "
+                    "with dimensionality ",
+                    arg_sample_dim, ")."));
+    const auto &schema = Operator<Backend>::GetSpec().GetSchema();
+    if (schema.ArgSupportsPerFrameInput(arg_name) && IsPerFrame(arg_tensor)) {
+      // Do not error out but simply ignore `F` layout of the argument input
+      // if it is not marked as per-frame in schema to be consistent with operators
+      // that do not support per-frame at all
+      return ExpandPerFrameLike(arg_tensor, arg_name, expand_idx, expand_desc);
+    }
+    return BroadcastLike(arg_tensor, expand_desc);
+  }
+
+  TensorVector<CPUBackend> ExpandPerFrameLike(const TensorVector<CPUBackend> &arg_tensor,
```

Comment:
Depending on the situation it may be `unfold_outer_dims` or broadcasting or mix of the two. The mix could be expressed as a composition of unfolding and broadcasting or broadcasting and unfolding depending on the FC or CF case, but is the creation of intermediate tensorvector worth it?
Or alternatively, I could try to get rid of `unfold_outer_dims` and make it a special case of unfold_broadcast mix, but it likely will be rather convoluted function.
To further complicate the picture we also may want to have multi-input cases with "FC" and "CF" input at the same time (or equivalently support for per-channel args) where we would need to transpose and unfold.

In general it seemed to me like having the "specializations" separated, they may be slightly repetitive but should be: 1. more efficient and 2. easier to read with clearer assumptions on the inputs and better error messaging and validation.

Reply:
I renamed the methods used for inputs, args and outputs expansion and placed them one by one at the end of the class but as we discussed, without some concessions argument expansion is never just unfolding.